### ISSUE TITLE
CGrenade causes crash on Restore under special conditions

### ISSUE THREAD

**Issue by BryanHaley** [SamVanheer/halflife-unified-sdk#352]
On semi-rare occurrences, the game will crash back to the main menu on trying to load with this vague error message in the console: 
`Host_Error: PF_precache_model_I: Bad string ''`

I tracked this down to `CGrenade::Precache`. On some instances, `pev->model` will point to 0, which will result in an invalid/empty string passed as an argument to `PrecacheModel`.

Here is a save that will crash: [quick.sav](https://drive.google.com/file/d/1P_VYsW8gn5DTf2qIxdHc_118lyQbS2sG/view?usp=sharing)

Using this temporary work-around hack will allow the save to load successfully:
```
void CGrenade::Precache()
{
	// Hack to avoid crash
	if (pev->model == 0)
		pev->model = MAKE_STRING("models/grenade.mdl");

	PrecacheModel(STRING(pev->model));
}
```
Obviously it would be more ideal to figure out why pev->model is null in order to do a more proper fix.

I haven't confirmed this yet, but it seems saving at the exact right time after the grenade explodes will trigger the issue. `CGrenade::Explode` sets `pev->model` to `iStringNull`. I'm guessing there's some frames between `CGrenade::Explode` being called and the entity being removed where the game can save and cause this issue.

**Comment by BryanHaley** [comment]
Added a candidate fix #353 . This one just adds a null check to `CGrenade::Precache`. Since `pev->model` being null is intended behavior after the grenade explodes, this might be appropriate.

Other ideas to fix:
- Don't write the `CGrenade` to the save file if `CGrenade::Explode` has been called
- Kill the `CGrenade` on the same frame it explodes

What fix is most appropriate might depend on how it affects gameplay and visuals when restoring a save. I haven't test what, if any, effect the alternate fix ideas might have; not sure if the explosion sprite and sound depend on the `CGrenade` still existing. If they don't, killing the `CGrenade` on the same frame it explodes might be the way to go. Candidate fix #353 works as expected (sprite and sound play as expected after load). Let me know what you think.

**Comment by BryanHaley** [comment]
Killing the `CGrenade` on the same frame it calls `CGrenade::Explode` probably isn't viable as it calls `CGrenade::Smoke` ~0.3 seconds after it explodes.

**Comment by BryanHaley** [comment]
Digging a little deeper, I was not able to reproduce this bug on base HLSDK 2.4. I'm not sure what's different, as `CGrenade::Explode` sets `pev->model` to `iStringNull` in 2.4 as well. It could possibly be related to strings being handled differently in HLSDK 2.4, or perhaps calling the engine's precache function directly has something to do with it. I don't particularly want to set up ancient Visual Studio to continue digging, but interesting nonetheless.

In any case, I added candidate fix #354 . This adds an empty string check and a null check to `UTIL_PrecacheModel` and throws a warning if it occurs. It may be more appropriate to simply let the game crash, but I think this provides a slightly more useful warning message. 

I think both #353 and #354 could work together, as it is probably unnecessary to throw a warning for `CGrenade::Precache` since setting `pev->model` to `iStringNull` while the entity still exists is intended behavior.

**Comment by SamVanheer** [comment]
> Digging a little deeper, I was not able to reproduce this bug on base HLSDK 2.4. I'm not sure what's different, as CGrenade::Explode sets pev->model to iStringNull in 2.4 as well.

That's because the original SDK always precaches the model even if `pev->model` is null.
